File: sdk/experimental/php/src/Output/Formatter/Formatter.php
<?php

declare(strict_types=1);

namespace HopTop\Kit\Output\Formatter;

/**
 * A Formatter encodes structured data to a writable stream.
 *
 * Implementations declare their key, file extensions, and the option keys
 * they accept. The Dispatcher validates --format-opt input against
 * options() before invoking render(), so render() may trust $opts to
 * contain only declared keys with values coerced to declared types.
 *
 * Mirrors py output.Formatter Protocol, ts Formatter<T> interface, and go
 * console/output.Formatter interface.
 */
interface Formatter
{
    /** Unique format identifier exposed via --format <key>. */
    public function key(): string;

    /**
     * File extensions (with leading dot, e.g. ".csv") that map to this
     * formatter for --output extension inference. May be empty.
     *
     * @return list<string>
     */
    public function extensions(): array;

    /**
     * Option specs accepted by this formatter via --format-opt key=value.
     *
     * @return list<OptionSpec>
     */
    public function options(): array;

    /**
     * Render $data to $writer.
     *
     * $cols arrives already resolved: the Dispatcher has collapsed the
     * user's --cols and the caller's ColumnSpec list into one ordered list
     * of column names, so a formatter never sees that precedence. Each name
     * is both the header label and the row lookup — header == key, so there
     * is no second mapping step. An empty $cols means "no column source
     * applied"; fall back to the payload's own key order.
     *
     * Zero rows emits nothing, not even a bare header row — emptiness is
     * decided by row count, never by column count, because $cols is
     * populated for row-less payloads too.
     *
     * @param resource              $writer fopen-style stream handle
     * @param mixed                 $data   single row or list of rows
     * @param array<string,mixed>   $opts   validated options
     * @param list<string>          $cols   resolved column projection ([] = "infer from payload")
     */
    public function render(mixed $writer, mixed $data, array $opts, array $cols): void;
}

File: sdk/experimental/rs/src/output/formatter.rs
//! Formatter trait.
//!
//! Mirrors py `Formatter` Protocol, ts `Formatter<T>` interface, go
//! `Formatter` interface, and php `Formatter` interface.

use std::io::Write;

use serde_json::Value;

use super::option::{OptionSpec, Options};

/// A Formatter encodes structured data to a writable stream.
///
/// Implementations declare their key, file extensions, and the option
/// keys they accept. The Dispatcher validates `--format-opt` input
/// against `options()` before invoking `render()`, so `render()` may
/// trust `opts` to contain only declared keys with values coerced to
/// declared types.
pub trait Formatter: Send + Sync {
    /// Unique format identifier exposed via `--format <key>`.
    fn key(&self) -> &'static str;

    /// File extensions (with leading dot, e.g. `".csv"`) that map to
    /// this formatter for `--output` extension inference. May be empty.
    fn extensions(&self) -> &'static [&'static str];

    /// Option specs accepted by this formatter via `--format-opt key=value`.
    fn options(&self) -> &'static [OptionSpec];

    /// Render `data` to `out`.
    ///
    /// `opts` contains only validated option values keyed by
    /// [`OptionSpec::name`]; missing keys mean the caller did not set
    /// them (render should fall back to spec defaults — which the
    /// Dispatcher already merges in via `parse_options`).
    ///
    /// `cols` is the user-requested column projection; an empty slice
    /// means "all default columns".
    fn render(
        &self,
        out: &mut dyn Write,
        data: &Value,
        opts: &Options,
        cols: &[String],
    ) -> std::io::Result<()>;
}
